src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,178 @@
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode
+                = new Node<>(null, value, null);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (index == 0) {
+            newNode.next = head;
+            if (head != null) {
+                head.prev = newNode;
+            }
+            head = newNode;
+            if (tail == null) {
+                tail = newNode;
+            }
+        } else if (index == size) {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        } else {
+            Node<T> current = findNodeByIndex(index - 1);
+            newNode.next = current.next;
+            newNode.prev = current;
+            current.next.prev = newNode;
+            current.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T newNodes : list) {
+            add(newNodes);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        Node<T> returnNode = findNodeByIndex(index);
+        return returnNode.data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        Node<T> updateNode = findNodeByIndex(index);
+        T oldNode = updateNode.data;
+        updateNode.data = value;
+        return oldNode;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        Node<T> current = null;
+        T nodeToRemove = null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+        if (size == 1) {
+            nodeToRemove = head.data;
+            head = null;
+            tail = null;
+            size--;
+            return nodeToRemove;
+        } else if (index == 0) {
+            nodeToRemove = head.data;
+            head = head.next;
+            head.prev = null;
+            size--;
+            return nodeToRemove;
+        } else if (index == size - 1) {
+            nodeToRemove = tail.data;
+            tail = tail.prev;
+            tail.next = null;
+            size--;
+            return nodeToRemove;
+        } else {
+            current = findNodeByIndex(index);
+            nodeToRemove = current.data;
+            unlink(current);
+        }
+        return nodeToRemove;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> nodeToRemove = head;
+        while (nodeToRemove != null) {
+            if ((object == null && nodeToRemove.data == null)
+                    || (object != null && object.equals(nodeToRemove.data))) {
```

Comment:
remove redundant parentheses 
```suggestion
            if (object == nodeToRemove.data
                    || object != null && object.equals(nodeToRemove.data)) {
```